examples/all-clusters-app/all-clusters-common/src/microwave-oven-mode.cpp

```diff
@@ -0,0 +1,106 @@
+
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::MicrowaveOvenMode;
+using chip::Protocols::InteractionModel::Status;
+template <typename T>
+using List              = chip::app::DataModel::List<T>;
+using ModeTagStructType = chip::app::Clusters::detail::Structs::ModeTagStruct::Type;
+
+static MicrowaveOvenModeDelegate * gMicrowaveOvenModeDelegate = nullptr;
+static ModeBase::Instance * gMicrowaveOvenModeInstance        = nullptr;
+
+CHIP_ERROR MicrowaveOvenModeDelegate::Init()
+{
+    return CHIP_NO_ERROR;
+}
+
+// todo refactor code by making a parent class for all ModeInstance classes to reduce flash usage.
+void MicrowaveOvenModeDelegate::HandleChangeToMode(uint8_t NewMode, ModeBase::Commands::ChangeToModeResponse::Type & response)
+{
+    response.status = to_underlying(ModeBase::StatusCode::kSuccess);
+}
+
+CHIP_ERROR MicrowaveOvenModeDelegate::GetModeLabelByIndex(uint8_t modeIndex, chip::MutableCharSpan & label)
+{
+    if (modeIndex >= ArraySize(kModeOptions))
+    {
+        return CHIP_ERROR_PROVIDER_LIST_EXHAUSTED;
+    }
+    return chip::CopyCharSpanToMutableCharSpan(kModeOptions[modeIndex].label, label);
+}
+
+CHIP_ERROR MicrowaveOvenModeDelegate::GetModeValueByIndex(uint8_t modeIndex, uint8_t & value)
+{
+    if (modeIndex >= ArraySize(kModeOptions))
+    {
+        return CHIP_ERROR_PROVIDER_LIST_EXHAUSTED;
+    }
+    value = kModeOptions[modeIndex].mode;
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR MicrowaveOvenModeDelegate::GetModeTagsByIndex(uint8_t modeIndex, List<ModeTagStructType> & tags)
+{
+    if (modeIndex >= ArraySize(kModeOptions))
+    {
+        return CHIP_ERROR_PROVIDER_LIST_EXHAUSTED;
+    }
+
+    if (tags.size() < kModeOptions[modeIndex].modeTags.size())
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    std::copy(kModeOptions[modeIndex].modeTags.begin(), kModeOptions[modeIndex].modeTags.end(), tags.begin());
+    tags.reduce_size(kModeOptions[modeIndex].modeTags.size());
+
+    return CHIP_NO_ERROR;
+}
+
+ModeBase::Instance * MicrowaveOvenMode::Instance()
+{
+    return gMicrowaveOvenModeInstance;
+}
+
+void MicrowaveOvenMode::Shutdown()
+{
+    if (gMicrowaveOvenModeInstance != nullptr)
+    {
+        delete gMicrowaveOvenModeInstance;
+        gMicrowaveOvenModeInstance = nullptr;
+    }
+    if (gMicrowaveOvenModeDelegate != nullptr)
+    {
+        delete gMicrowaveOvenModeDelegate;
+        gMicrowaveOvenModeDelegate = nullptr;
+    }
+}
+
+void emberAfMicrowaveOvenModeClusterInitCallback(chip::EndpointId endpointId)
+{
+    VerifyOrDie(endpointId == 1); // this cluster is only enabled for endpoint 1.
+    VerifyOrDie(gMicrowaveOvenModeDelegate == nullptr && gMicrowaveOvenModeInstance == nullptr);
+    gMicrowaveOvenModeDelegate = new MicrowaveOvenMode::MicrowaveOvenModeDelegate;
+    gMicrowaveOvenModeInstance =
+        new ModeBase::Instance(gMicrowaveOvenModeDelegate, 0x1, MicrowaveOvenMode::Id, chip::to_underlying(Feature::kOnOff));
```

Comment:
It should be possible to externally set/inject these delegates. Otherwise it's impossible to have different microwave mode implementations without forking the SDK. Please allow external setters and DO NOT allocate the delegate here.